### ISSUE TITLE
Return dependencies in optimization calls

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#2577]
A non-trivial fraction of our scheduling overhead comes from calculating dependencies of the task graph.  We currently do this redundantly in a few places.  A while ago we changed many of the optimization functions to return both a task graph and the dependencies so that this can be passed along to other optimizations.  However, the full `dd.optimize` or `da.optimize` functions don't do this, they only return the task graph.

I am inclined to change this, so that top-level optimization functions also return dependencies.  This will help us to reduce overhead costs.  However, this also is a break to top-level API and will break user code from some more advanced users and so perhaps warrants some discussion.

cc @shoyer @jcrist

**Comment by jcrist** [comment]
Fine by me. This also could allow custom mutable mappings to generate their dependencies in a potentially more efficient way. I'm curious, on real problems how much of the time is saved by being smarter about this?

>  However, this also is a break to top-level API and will break user code from some more advanced users.

Users writing their own optimizations/creating custom classes are already using fringe features. Could either do a clean break (easy for us, hard for users) or deprecate with something like:

```
res = x._optimize(x.dask, x._keys())
if len(res) == 2:
    dsk, dependencies = res
else:
    warn("In the future optimize methods should return ""
         "dependencies.")
    dsk = res
    dependencies = ...
```

**Comment by mrocklin** [comment]
> I'm curious, on real problems how much of the time is saved by being smarter about this?

Yeah, that's a valid question.  This is a centralized scheduling cost, and so becomes an issue as we scale to larger distributed problems, such as occur for large-scale array analysis.  Today interaction times on large datasets can be slowed down quite a bit at the graph optimization and submission stage.  People sometimes wait a minute or so between hitting persist/compute and seeing anything show up on the scheduler.

**Comment by shoyer** [comment]
This seems reasonable to me.

We don't use the optimization API directly in xarray yet, so there's no concern there.
